src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs

```diff
@@ -784,6 +810,192 @@ private void PopulateEncLogTableParameters()
+                    rowId++;
+                    hasData = enumerator.MoveNext();
+                }
+
+                if (emittedIndex == emittedAttributeParents.Count)
+                {
+                    break;
+                }
+            }
+
+            // If we exhausted the original custom attributes table but there are still attributes that were emitted
+            // they must be part of the deltas
+            for (int i = emittedIndex; i < emittedAttributeParents.Count; i++)
+            {
+                skippedParents.Add(emittedAttributeParents[i]);
+            }
+
+            return skippedParents;
+        }
+
+        private void PopulateEncLogTableCustomAttributesFromDeltas(List<EntityHandle> remainingEmittedAttributeParents, int[] encLogRows)
+        {
+            if (remainingEmittedAttributeParents.Count == 0)
+            {
+                return;
+            }
+
+            // The data in _previousGeneration.CustomAttributesAdded is not nicely sorted, or even necessarily contiguous
+            // so we need to map each target onto the rows its attributes occupy so we know which rows to update
+            var attributeMap = CreateExistingAttributeMap(_previousGeneration.OriginalMetadata.MetadataReader, _previousGeneration.CustomAttributesAdded, out var lastRowId);
+
+            int rowId;
+            var nextRowId = lastRowId + 1;
+            var logIndex = 0;
+            foreach (var customAttributeTarget in remainingEmittedAttributeParents)
+            {
+                if (attributeMap.TryGetValue(customAttributeTarget, out var queue) &&
+                    queue.Count > 0)
+                {
+                    // If there is already an attribute for this parent, then pull the next available row number
+                    // off the queue and update it
+                    rowId = queue.Dequeue();
+                }
+                else
+                {
+                    // Otherwise, either this is a new parent that hasn't had an attribute before, or we've run
+                    // out of existing rows to update. Either way, we want to add to the end of the table.
+                    rowId = nextRowId++;
+
+                    // Keep track of the addition for passing to the next generation
+                    _customAttributesAdded[rowId] = customAttributeTarget;
+                }
+
+                // remainingEmittedAttributeParents only contains records that haven't been logged, but since we know it is in the
+                // right order, we don't need to know what the original index we can just fill log rows in order, but and skip any
+                // that were already filled in the previous phase.
+                Debug.Assert(logIndex < encLogRows.Length, "encLogRows cannot be larger than remainingEmittedAttributeParents.");
+                while (encLogRows[logIndex] != 0)
```

Comment:
>while (encLogRows[logIndex] != 0) [](http://example.com/codeflow?start=16&length=33)

How do we know `encLogRows` contains a least one more `0`?

Reply:
Yeah, good question, this could definitely do with more asserts.